### DISCUSSION TITLE
`has_many` in NotificationExtentions does not work

### DISCUSSION THREAD

**Discussion by bvogel** [excid3/noticed#474] [Q&A]
Hi,

still in the process of refactoring our code for v2 I found another hiccup I can't resolve. we have this as extension:
```
module NotificationExtensions
  extend ActiveSupport::Concern

  included do
    has_many :notification_relations, dependent: :destroy, inverse_of: :noticed_notification

    accepts_nested_attributes_for :notification_relations,
                                  reject_if: proc { |attributes| attributes["target_id"].blank? },
                                  allow_destroy: true
  end
end
```

with NotificationRelation being:
```
class NotificationRelation < ApplicationRecord
  belongs_to :noticed_notification, inverse_of: :notification_relations, class_name: "Noticed::Notification"
  belongs_to :target, polymorphic: true
end
```

I tried overwriting `recipient_attributes_for` from `Noticed::Deliverable`
```
    def recipient_attributes_for(recipient)
      attributes = {
        type: "#{self.class.name}::Notification",
        recipient_type: recipient.class.base_class.name,
        recipient_id: recipient.id
      }
      return attributes if params[:relations].blank?

      attributes[:notification_relations_attributes] =
        params[:relations].map { |relation| { target_id: relation.id, target_type: relation.class.name } }

      attributes
    end
```
but I get `ActiveModel::UnknownAttributeError: unknown attribute 'notification_relations_attributes' for Noticed::Notification` which seems to be related to `notifications.insert_all!(recipients_attributes, record_timestamps: true) if recipients_attributes.any?`  from `Noticed::Deliverable`.

The only way I got this to work was  overwriting the complete `def deliver(recipients = nil, enqueue_job: true, **options)` method and using this block
```
          if recipients_attributes.any?
            recipients_attributes.each do |attributes|
              notifications.create!(attributes)
            end
          end
```

This seems not to be a correct approach, so I'd love to hear how I'd go about this. (I briefly started fiddling around with the undocumented `def autosave_associated_records_for_*` method, but that wasn't even picked up in my tests.)

**Comment by excid3** [comment]
We use `insert_all` which isn't going to instantiate ActiveRecord objects like normal inserts to support large bulk deliveries. If you're not doing that, your overrides will work.

**Reply by bvogel** [reply]
so again we are running in a non-standard use-case? This will now mean to keep tracking changes to the `Deliverable` class for future updates. Will there be a chance that callbacks or extendable methods will make this easier to maintain?

BTW: Thank you SOOO MUCH for being so responsive and understanding.

**Reply by excid3** [reply]
Callbacks and such won't run with `insert_all`. See: https://api.rubyonrails.org/v7.1.3.4/classes/ActiveRecord/Persistence/ClassMethods.html#method-i-insert_all

**Reply by bvogel** [reply]
Sorry, I didn't mean callbacks on `insert_all` but probably have a method `def insert_notifications` and have that make the insert_all call so it would be easier to overwrite. 

The callbacks would be something like `before_insert_notifications(recipient_attributes)` and `after_insert_notifications(notifications)` that could be used to handle custom code to add additional fields or do relations to the notifications - just a thought...

These are probably two ways archiving the same thing?
